daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -194,132 +258,161 @@ private[lf] object Anf {
+  private[this] def transformExp[A](
+      depth: DepthA,
+      env: Env,
+      exp: SExpr,
+      transform: Tx[SExpr, A],
+      k: K[AExpr, A]): Trampoline[A] =
+    exp match {
+      case atom0: SExprAtomic =>
+        val atom = makeRelativeA(depth)(makeAbsoluteA(env, atom0))
+        Bounce(() => transform(depth, atom, k))
+
+      case x: SEVal => Bounce(() => transform(depth, x, k))
+      case x: SEImportValue => Bounce(() => transform(depth, x, k))
+
+      case SEAppGeneral(func, args) => {
+        def tx2(
+            func: AbsAtom)(depth: DepthA, args: List[AbsAtom], txK: K[AExpr, A]): Trampoline[A] = {
+          val func1 = makeRelativeA(depth)(func)
+          val args1 = args.map(makeRelativeA(depth))
+          Bounce(() => transform(depth, SEAppAtomic(func1, args1.toArray), txK))
+        }
+        def tx1(depth: DepthA, func: AbsAtom, txK1: K[AExpr, A]): Trampoline[A] = {
+          Bounce(() => atomizeExps(depth, env, args.toList, tx2(func), txK1))
+        }
+        Bounce(() => atomizeExp(depth, env, func, tx1, k))
+      }
+      case SEMakeClo(fvs0, arity, body0) =>
+        val fvs = fvs0.map((loc) => makeRelativeL(depth)(makeAbsoluteL(env, loc)))
+        val body = flattenToAnf(body0).wrapped
+        Bounce(() => transform(depth, SEMakeClo(fvs, arity, body), k))
+
+      case SECase(scrut, alts0) => {
+        def tx(depth: DepthA, scrut: AbsAtom, txK: K[AExpr, A]): Trampoline[A] = {
+          val scrut1 = makeRelativeA(depth)(scrut)
+          Bounce(
+            () =>
+              flattenAlts(
+                depth,
+                env,
+                alts0,
+                alts => Bounce(() => transform(depth, SECaseAtomic(scrut1, alts), txK))))
+        }
+        Bounce(() => atomizeExp(depth, env, scrut, tx, k))
+      }
+
+      case SELet(rhss, body) =>
+        val expanded = expandMultiLet(rhss.toList, body)
+        Bounce(() => transformExp(depth, env, expanded, transform, k))
+
+      case SELet1General(rhs, body) =>
+        Bounce(() => transformLet1(depth, env, rhs, body, transform, k))
+
+      case SECatch(body0, handler0, fin0) =>
+        Bounce(
+          () =>
+            flattenExp(
+              depth,
+              env,
+              body0,
+              body =>
+                Bounce(() =>
+                  flattenExp(
+                    depth,
+                    env,
+                    handler0,
+                    handler =>
+                      Bounce(
+                        () =>
+                          flattenExp(
+                            depth,
+                            env,
+                            fin0,
+                            fin =>
+                              Bounce(() =>
+                                transform(
+                                  depth,
+                                  SECatch(body.wrapped, handler.wrapped, fin.wrapped),
+                                  k))))
+                ))
+          ))
```

Comment:
There's probably a monad in here somewhere, but I don't think you should worry about finding it.